### PR TITLE
:warning: Fakeclient: reject create/update if Name isn't provided

### LINKED ISSUE #838: fakeclient should reject create|update without meta.name
Re: https://github.com/kubernetes-sigs/controller-runtime/pull/832#discussion_r385186769

Adding some of the basic validation you would find in other clients (ie. [client-go](https://github.com/kubernetes/client-go/blob/master/dynamic/simple.go#L136-L139)) would be nice to have.

### PR DESCRIPTION
**Pull request by GrigoriyMikhalkin**
Fixes #838

### CODE REVIEW

**Comment by alvaroaleman** [comment]
Thanks! Could you change the title prefix to `:warning:` so that ppl are aware that this is a breaking change, as it will break all tests that do not provide a name?

**Comment by GrigoriyMikhalkin** [comment]
> Thanks! Could you change the title prefix to `:warning:` so that ppl are aware that this is a breaking change, as it will break all tests that do not provide a name?

Done

File: pkg/client/fake/client.go [hunk 1]
```diff
@@ -85,6 +85,9 @@ func (t versionedTracker) Create(gvr schema.GroupVersionResource, obj runtime.Ob
 	if err != nil {
 		return err
 	}
+	if accessor.GetName() == "" {
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:88]: This needs to be a `NewInvalid`, this is how that looks like from a real api server:
```
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "ConfigMap \"\" is invalid: metadata.name: Required value: name or generateName is required",
                  Reason: "Invalid",
                  Details: {
                      Name: "",
                      Group: "",
                      Kind: "ConfigMap",
                      UID: "",
                      Causes: [
                          {
                              Type: "FieldValueRequired",
                              Message: "Required value: name or generateName is required",
                              Field: "metadata.name",
                          },
                      ],
                      RetryAfterSeconds: 0,
                  },
                  Code: 422,
              }
```

**Pull request merged**